### ISSUE TITLE
event.currentTarget not working when using mount

### ISSUE THREAD

**Issue by kunyan** [airbnb/enzyme#1943]
Thanks for reporting an issue to us! We're glad you are using and invested in Enzyme.
Before submitting, please read over our commonly reported issues to prevent duplicates!

### All common issues

* [common issues](../blob/master/docs/common-issues.md)

### Notoriously common issues

* [Webpack build issues](../blob/master/docs/common-issues.md#webpack-build-issues)
* [Cannot find module 'react-dom/lib/ReactTestUtils'](../blob/master/docs/common-issues.md#error-cannot-find-module-react-domlibreacttestutils)
* [Query Selector fails](../blob/master/docs/common-issues.md#query-selector-fails)
* [Testing third party libraries](../blob/master/docs/common-issues.md#testing-third-party-libraries)

If you haven't found any duplicated issues, please report it with your environment!

### Current behavior

`currentTarget` not working when using `mount`
```javascript
const wrapper = mount(<input type="text" onChange={(e)=> console.log(e.currentTarget.value)}/>)

wrapper.simulate('change', {currentTarget: {
  value: 'Hello'
}});
```

```
```

`currentTarget` working when using `shallow`
```javascript
const wrapper = shallow(<input type="text" onChange={(e)=> console.log(e.currentTarget.value)}/>)

wrapper.simulate('change', {currentTarget: {
  value: 'Hello'
}});
```

```
Hello
```
### Expected behavior

```javascript
const wrapper = mount(<input type="text" onChange={(e)=> console.log(e.currentTarget.value)}/>)

wrapper.simulate('change', {currentTarget: {
  value: 'Hello'
}});
```

```
Hello
```

#### API

- [x] shallow
- [x] mount
- [ ] render

#### Version

| library             | version
| ------------------- | -------
| enzyme              | 3.1.15
| react               |  16.6.3
| react-dom           |  16.6.3
| react-test-renderer | 16.6.3
| adapter (below)     |  1.0.3

#### Adapter

- [x] enzyme-adapter-react-16
- [ ] enzyme-adapter-react-16.3
- [ ] enzyme-adapter-react-16.2
- [ ] enzyme-adapter-react-16.1
- [ ] enzyme-adapter-react-15
- [ ] enzyme-adapter-react-15.4
- [ ] enzyme-adapter-react-14
- [ ] enzyme-adapter-react-13
- [ ] enzyme-adapter-react-helper
- [ ] others ( )

**Comment by ljharb** [comment]
I would suggest avoiding `.simulate` - it doesn't faithfully simulate anything. Instead, manually invoke the desired prop - ie, `wrapper.prop('onChange')(fakeEvent)`.

**Comment by Totemika** [comment]
@ljharb's solution worked for me but the documentation says that any object you pass to the `simulate` method as a second parameter, will be passed to the event handler, and that seems to have been true for a long time, since many [Stack Overflow]() answers are using it. But the truth is that it's not currently working when using `mount` but it *does work* with `shallow`. 
Maybe we should update the documentation to warn about this? If not, isn't it a bug?

![image](https://user-images.githubusercontent.com/4906291/50019326-2b50ea80-ffca-11e8-9a5f-81f64bc6dcc9.png)

On the other hand @ljharb has been advising people to use `wrapper.prop('eventName')` instead of `.simulate` for a long time. What is the use of `simulate` then? 

Thank you

**Comment by ljharb** [comment]
`simulate` only exists in enzyme because it was present in `react-test-renderer`, and we added it to enzyme before realizing that it's badly named and unreliably implemented.

I do agree that if the documentation is wrong, and/or if mount/shallow are inconsistent without good reason, then that's something that should be fixed. @Totemika can you file a new issue about that part, with repro code (showing the difference between shallow and mount)?

**Comment by steveHornseyDeveloper** [comment]
Why does simulate behave differently in mount and shallow? I am writing an integration test and I'd rather not trigger an event in an artificial way like by triggering an onChange prop, as I want to keep the test as similar to the browser as possible.

**Comment by ljharb** [comment]
@steveHornseyDeveloper triggering the onChange prop is the proper way to do it; if you're not using something like webdriver to make an *actual* click, then you're not getting any value out of implicitly calling `onChange` by using simulate.

**Comment by wyqydsyq** [comment]
> @steveHornseyDeveloper triggering the onChange prop is the proper way to do it; if you're not using something like webdriver to make an _actual_ click, then you're not getting any value out of implicitly calling `onChange` by using simulate.

Perhaps if triggering the desired handler from the mounted component's props is the "proper" way to do it, then this fact should be documented somewhere?

[Searching Google for "enzyme trigger event"](https://www.google.com/search?sxsrf=ACYBGNTGBw0mWOnqYyN01XTnbDl-Csky9w%3A1577681571061&ei=o4IJXsK1Asuq0PEP0LO-4Ak&q=enzyme+trigger+event&oq=enzyme+trigger+event&gs_l=psy-ab.3..0j0i22i30l4.1453.5726..5835...0.3..0.239.3828.2-17......0....1..gws-wiz.......0i71j0i7i30j0i13j35i39j0i273j0i67j0i131i20i263j0i20i263.yDl0JsAZfEw&ved=0ahUKEwiCk_HsydzmAhVLFTQIHdCZD5wQ4dUDCAs&uact=5) gives results where every link on the first page tells you to use `.simulate()`, the first two of which are links to the official Enzyme docs.

Further, all the examples of triggering events in Enzyme's official docs suggest to use `.simulate()`. 

Even just mentioning what you've said in this issue thread under the [https://airbnb.io/enzyme/docs/api/ShallowWrapper/simulate.html#common-gotchas](common gotchas) of this function would be a lot clearer than users having to dig through threads to find this information.

**Comment by ljharb** [comment]
A PR to improve the docs in the way you suggest would be appreciated.

**Comment by wyqydsyq** [comment]
Apologies if my last comment came off as abrasive, I was trying to express bewilderment that the *correct* way to do something as essential as testing event behaviour isn't really documented

I'd be happy to submit a PR, but what I'm thinking though is if this is really the recommended way to trigger events on components you're testing in Enzyme, there should probably be a dedicated guide on testing events using this method or at least a section on the [.props()](https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/props.md) page showing it used for that rather than just a footnote on the `.simulate()` page, because there's a lot of misinformation on the web instructing people to use `.simulate()` and when reading the docs it looks like that's the de-facto method to test event handlers on your components, so a more extensive and visible correction to that might help a lot more.

If that would be a welcome change I'm happy to have crack at writing it, otherwise I'll just write an added gotcha on the `.simulate()` page.

**Comment by mickeypuri** [comment]
@kunyan yes, this tripped me up too. I had to use `event.target` to get it working with `mount`.

However I found if you try and use `event.target` with `shallow`, then it fails. 

So you need   `event.currentTarget` to work with `shallow `
and `event.target` to work with `mount `

**Comment by chriswininger** [comment]
fyi wrapper.prop('onChange')(fakeEvent) does seem to work but you get lots of angry red text

```
  console.error node_modules/react-dom/cjs/react-dom.development.js:558
    Warning: An update to ComponentWithNxTextInput inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
        in ComponentWithNxTextInput (created by WrapperComponent)
        in WrapperComponent
```

What is the right way to do this?! This seems like a fairly basic thing to be broken out of the box.

**Comment by chriswininger** [comment]
I found an approach that seems to work in this thread: https://github.com/enzymejs/enzyme/issues/218

```
const mountedComponent = mount(<ComponentWithNxTextInput />);

const input = mountedComponent.find('input');

input.getDOMNode().setAttribute('value', givenInputValue);
input.simulate('change', { currentTarget: input });
```

The key is pointing currentTarget to a real dom node reference. This seems to work better than props('onChange') avoiding all that ugly red text I mentioned.
